IRQ_CONNECT and IRQ_DIRECT_CONNECT throw compile error with CONFIG_CPLUSPLUS

**Describe the bug**
When compiling using C++ support, the Z_ISR_DELCARE macro does not compile.
```
In file included from zephyr/include/arch/arm/aarch32/irq.h:19,
                 from zephyr/include/arch/arm/aarch32/arch.h:27,
                 from zephyr/include/arch/cpu.h:19,
                 from zephyr/include/kernel_includes.h:34,
                 from zephyr/include/kernel.h:17,
                 from zephyr/include/zephyr.h:18,
                 from ../src/main.cpp:1:
../src/main.cpp: In function 'void main()':
zephyr/include/sw_isr_table.h:70:17: error: invalid conversion from 'void (*)(void*)' to 'void*' [-fpermissive]
   70 |    {irq, flags, &func, (void *)param}
zephyr/include/arch/arm/aarch32/irq.h:82:2: note: in expansion of macro 'Z_ISR_DECLARE'
   82 |  Z_ISR_DECLARE(irq_p, 0, isr_p, isr_param_p); \
      |  ^~~~~~~~~~~~~
zephyr/include/irq.h:51:2: note: in expansion of macro 'ARCH_IRQ_CONNECT'
   51 |  ARCH_IRQ_CONNECT(irq_p, priority_p, isr_p, isr_param_p, flags_p)
      |  ^~~~~~~~~~~~~~~~
../src/main.cpp:9:3: note: in expansion of macro 'IRQ_CONNECT'
    9 |   IRQ_CONNECT(TIMER0_IRQn, 0, responder_irq_handler, NULL, 0);
      |   ^~~~~~~~~~~
```
```
In file included from zephyr/include/arch/arm/aarch32/irq.h:19,
                 from zephyr/include/arch/arm/aarch32/arch.h:27,
                 from zephyr/include/arch/cpu.h:19,
                 from zephyr/include/kernel_includes.h:34,
                 from zephyr/include/kernel.h:17,
                 from zephyr/include/zephyr.h:18,
                 from ../src/main1.cpp:1:
../src/main1.cpp: In function 'void main()':
zephyr/include/sw_isr_table.h:70:17: error: invalid conversion from 'void (*)()' to 'void*' [-fpermissive]
   70 |    {irq, flags, &func, (void *)param}
zephyr/include/arch/arm/aarch32/irq.h:89:2: note: in expansion of macro 'Z_ISR_DECLARE'
   89 |  Z_ISR_DECLARE(irq_p, ISR_FLAG_DIRECT, isr_p, NULL); \
      |  ^~~~~~~~~~~~~
zephyr/include/irq.h:116:2: note: in expansion of macro 'ARCH_IRQ_DIRECT_CONNECT'
  116 |  ARCH_IRQ_DIRECT_CONNECT(irq_p, priority_p, isr_p, flags_p)
      |  ^~~~~~~~~~~~~~~~~~~~~~~
../src/main1.cpp:11:3: note: in expansion of macro 'IRQ_DIRECT_CONNECT'
   11 |   IRQ_DIRECT_CONNECT(TIMER0_IRQn, 0, responder_irq_handler, 0);
      |   ^~~~~~~~~~~~~~~~~~
```
After converting my project back to C and recompiling, the errors go away and it compiles and runs perfectly.

**To Reproduce**
Use either of the main files at the bottom of the document as the sole source file.
CONFIG_CPLUSPLUS=y
BOARD nrf52840_pca10056

**Impact**
This is a pretty hefty bug for us because we are unable to meet the timing requirements of our application without being able to utilize interrupts. It seems like a major bug if a real-time OS can't support declaring interrupts.


**Environment:**
 - OS: Ubuntu 18.04
 - Toolchain gnuarmemb (gcc-arm-none-eabi-9-2019-q4-major)
 - Commit SHA 616c7d5becd4e89870830e586f2626758653b1ea (v2.2.0)

**main.cpp**
```
#include <zephyr.h>

void responder_irq_handler(void *args)
{
}

void main(void)
{
  IRQ_CONNECT(TIMER0_IRQn, 0, responder_irq_handler, NULL, 0);
  irq_enable(TIMER0_IRQn);
}

```
**main1.cpp**
```
#include <zephyr.h>

ISR_DIRECT_DECLARE(responder_irq_handler)
{
  ISR_DIRECT_PM();
  return 1;
}

void main(void)
{
  IRQ_DIRECT_CONNECT(TIMER0_IRQn, 0, responder_irq_handler, 0);
  irq_enable(TIMER0_IRQn);
}

```

--- a/include/sw_isr_table.h
+++ b/include/sw_isr_table.h
@@ -67,7 +67,7 @@ struct _isr_list {
 #define Z_ISR_DECLARE(irq, flags, func, param) \
 	static Z_DECL_ALIGN(struct _isr_list) Z_GENERIC_SECTION(.intList) \
 		__used _MK_ISR_NAME(func, __COUNTER__) = \
-			{irq, flags, &func, (void *)param}
+			{irq, flags, (void *)&func, (void *)param}
 
 #define IRQ_TABLE_SIZE (CONFIG_NUM_IRQS - CONFIG_GEN_IRQ_START_VECTOR)
 

